includes/event-date.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace Wporg\TranslationEvents;
+
+use DateTime;
+use DateTimeImmutable;
+use DateTimeZone;
+use Exception;
+
+abstract class Event_Date extends DateTimeImmutable {
+	abstract public function is_end_date(): bool;
+	public $timezone;
+	public function __construct( string $date, DateTimeZone $timezone = null ) {
+		if ( ! $timezone ) {
+			$timezone = new DateTimeZone( 'UTC' );
+		}
+
+		try {
+			$utc_date = new DateTime( $date, new DateTimeZone( 'UTC' ) );
+			$utc_date->setTimezone( $timezone );
+		} catch ( Exception $e ) {
+			$utc_date = new DateTime( 'now', new DateTimeZone( 'UTC' ) );
+		}
+
+		parent::__construct( $utc_date->format( 'Y-m-d H:i:s' ), $timezone );
+		$this->timezone = $timezone;
+	}
+
+	/**
+	 * Get the standard formatted text for the date in UTC.
+	 *
+	 * @return string The date text.
+	 */
+	public function __toString(): string {
+		return $this->utc( 'Y-m-d H:i:s' );
+	}
+	/**
+	 * Get the local formatted text for the date in UTC.
+	 *
+	 * @return string The date text.
+	 */
+	public function utc( $format ): string {
+		return $this->setTimeZone( new DateTimeZone( 'UTC' ) )->format( $format );
+	}
+
+	/**
+	 * Generate variable text depending on when the event starts or ends.
+	 *
+	 * @return string The end date text.
+	 */
+	public function get_variable_text(): string {
```

Comment:
Since this is about presentation (displaying to the user), I think that this function could be moved out of this class into `templates/`, either directly in `events-list.php` (as it's currently only being called there), or if we expect it to be called from elsewhere in the future, in `helper-functions.php`.

Doing this would remove the need to have the `is_end_date()` method in the `Event_Date` class. It feels a bit weird that a start date would know whether it's an end date (it's a start date, so it should not be possible for it to be an end date), and moving the `get_variable_text()` out of this class would remove the need for the `is_end_date()` function, I think.

This would also probably mean that there is no longer the need to have both an `Event_Start_Date` and a `Event_End_Date` classes, there would only be an `Event_Date`.

To clarify, when we call `get_variable_text()`, we know whether we're calling it on an end date or not (`$event_end->get_variable_text()`), so `get_variable_text()` could take a boolean as argument that tells whether it's an end date or not: `get_variable_text( bool $is_end_date )`.